src/wasm-validator.h

```diff
@@ -46,13 +46,15 @@
 namespace wasm {
 
 // Print anything that can be streamed to an ostream
-template <typename T>
+  template <typename T,
+    typename std::enable_if<
+      !std::is_base_of<Expression, typename std::remove_pointer<T>::type>::value
```

Comment:
Well, I'm hoping with a larger refactoring we could remove the need for all of this printing code. Perhaps we can leave it for another PR. But iiuc, the issue here is to pass the option of printing the types, then we could have an API like this:
````
std::cerr << FullTypePrinter(expr)
````
which would print it with the full type. And so we could end up with just `cerr << [..]` for everything, and without these new templates and function calls?